academic_observatory_workflows/workflows/dqc_workflow.py

```diff
@@ -0,0 +1,456 @@
+        """Create the DataQualityCheck Workflow.
+
+        This workflow creates metadata for all the tables defined in the "datasets" dictionary.
+        If a table has already been checked before, it will no do it again. This based on if the
+        number of columns, rows or bytesof data stored in the table changes. We cannot use the "date modified"
+        from the Biguqery table object because it changes if any other metadata is modified, i.e. description, etc.
+
+        :param dag_id: the DAG ID.
+        :param cloud_workspace: the cloud workspace settings.
+        :param datasets: A dictionary of datasets holding tables that will processed by this workflow.
+        :param sensor_dag_ids: List of dags that this workflow will wait to finish before running.
+        :param bq_dataset_id: The dataset_id of where the dqc records will be stored.
+        :param bq_dataset_description: Description of the data quality check dataset.
+        :param bq_table_id: The name of the table in Bigquery.
+        :param bq_table_description: The description of the table in Biguqery.
+        :param schema_path: The path to the schema file for the records produced by this workflow.
+        :param start_date: The start date of the workflow.
+        :param schedule: Schedule of how often the workflow runs.
+        :param queue: Which queue this workflow will run.
+        """
+
+        super().__init__(
+            dag_id=dag_id,
+            start_date=start_date,
+            schedule=schedule,
+            tags=[Tag.data_quality_check],
+            queue=queue,
+        )
+
+        self.cloud_workspace = cloud_workspace
+        self.project_id = cloud_workspace.project_id
+        self.bq_dataset_id = bq_dataset_id
+        self.bq_table_id = bq_table_id
+        self.bq_dataset_description = bq_dataset_description
+        self.bq_table_description = bq_table_description
+        self.data_location = cloud_workspace.data_location
+        self.schema_path = schema_path
+        self.datasets = datasets
+
+        # Full table id for the DQC records.
+        self.dqc_full_table_id = f"{self.project_id}.{self.bq_dataset_id}.{bq_table_id}"
+
+        # If no sensor workflow is given, then it will run on a regular scheduled basis.
+        self.sensor_dag_ids = sensor_dag_ids
+
+        assert datasets, "No dataset or tables given for this DQC Workflow! Please revise the config file."
+
+    def make_dag(self) -> DAG:
+        """Create a DAG object for the workflow explicitly defining the tasks and task groups.
+
+        :return: The DAG object for the workflow."""
+
+        with self.dag:
+            # Create a group of sensors for the workflow.
+            task_sensor_group = []
+            if self.sensor_dag_ids:
+                with TaskGroup(group_id="dag_sensors") as tg_sensors:
+                    task_sensors = []
+
+                    for ext_dag_id in self.sensor_dag_ids:
+                        sensor = DagRunSensor(
```

Comment:
It would be good to change this to use the Airflow `ExternalTaskSensor` which should be more reliable than the DagRunSensor. Some of its parameters may need configuring so that it waits for the given DAG.

https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/sensors/external_task/index.html